File: content/snippets/go/interview_lcci_03_02_min_stack.go
package main

import "fmt"

type MinStack struct {
	values []int
	mins   []int
}

func (m *MinStack) Push(x int) {
	m.values = append(m.values, x)
	if len(m.mins) == 0 || x <= m.mins[len(m.mins)-1] {
		m.mins = append(m.mins, x)
	}
}

func (m *MinStack) Pop() {
	if len(m.values) == 0 {
		return
	}
	top := m.values[len(m.values)-1]
	if top == m.mins[len(m.mins)-1] {
		m.mins = m.mins[:len(m.mins)-1]
	}
	m.values = m.values[:len(m.values)-1]
}

func (m *MinStack) Top() int {
	if len(m.values) == 0 {
		return -1
	}
	return m.values[len(m.values)-1]
}

func (m *MinStack) GetMin() int {
	if len(m.mins) == 0 {
		return -1
	}
	return m.mins[len(m.mins)-1]
}

func main() {
	st := &MinStack{}
	st.Push(3)
	st.Push(5)
	st.Push(2)
	st.Push(2)
	fmt.Println(st.GetMin())
	st.Pop()
	fmt.Println(st.GetMin())
	st.Pop()
	fmt.Println(st.GetMin())
}

File: content/snippets/typescript/interview_lcci_03_02_min_stack.ts
class MinStack {
  private readonly values: number[] = [];
  private readonly mins: number[] = [];

  push(x: number): void {
    this.values.push(x);
    if (this.mins.length === 0 || x <= this.mins[this.mins.length - 1]) {
      this.mins.push(x);
    }
  }

  pop(): void {
    if (this.values.length === 0) {
      return;
    }
    const top = this.values.pop()!;
    if (top === this.mins[this.mins.length - 1]) {
      this.mins.pop();
    }
  }

  top(): number {
    return this.values.length ? this.values[this.values.length - 1] : -1;
  }

  getMin(): number {
    return this.mins.length ? this.mins[this.mins.length - 1] : -1;
  }
}

function main(): void {
  const st = new MinStack();
  st.push(3);
  st.push(5);
  st.push(2);
  st.push(2);
  console.log(st.getMin());
  st.pop();
  console.log(st.getMin());
  st.pop();
  console.log(st.getMin());
}

main();
